contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,244 @@
+        IERC20(WETH).approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+
+    /* ============ External Functions ============ */
+
+    receive() external payable {}
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming an index. This function must only be called before the first time
+     * this smart contract is used on any particular index, or when a new token is added
+     * to the index.
+     *
+     * @param _index    Address of the index being initialized
+     */
+    function initApprovals(address _index) external {
+        ISetToken.Position[] memory positions = ISetToken(_index).getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            IERC20 token = IERC20(positions[i].component);
+            token.approve(address(uniRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(basicIssuanceModule), PreciseUnitMath.maxUint256());
+        }
+    }
+
+    /**
+     * Redeems an Index and sells the underlying tokens using Uniswap
+     * or Sushiswap.
+     *
+     * @param _index        Address of the index being redeemed
+     * @param _amount       The amount of the index to redeem
+     * @param _isOutputETH  Set to true if the output token is Ether
+     * @param _outputToken  Address of output token. Ignored if _isOutputETH is true
+     */
+    function exchangeRedeem(address _index, uint256 _amount, bool _isOutputETH, address _outputToken) external {
+        ISetToken setToken = ISetToken(_index);
+        setToken.transferFrom(msg.sender, address(this), _amount);
+        basicIssuanceModule.redeem(ISetToken(_index), _amount, address(this));
+        liquidateSetPositions(_index);
+        if(_isOutputETH) {
+            msg.sender.transfer(address(this).balance);
+        } else {
+            purchaseTokenExactEther(_outputToken, address(this).balance);
+        }
+        emit ExchangeRedeem(msg.sender, _index, _isOutputETH ? address(0) : _outputToken, _amount);
+    }
+
+    /**
+     * Issues an Index by using swapping for the underlying tokens on Uniswap
+     * or Sushiswap. msg.value must be equal to the maximum price in ETH that you are
+     * willing to pay. Excess ETH is refunded.
+     *
+     * @param _index        Address of the index being issued
+     * @param _amount       Amount of the index to issue
+     * @param _isInputETH   Set to true if the input token is Ether
+     * @param _inputToken   Address of input token. Ignored if _isInputETH is true
+     * @param _maxSpend     Max erc20 balance to spend on issuing. Ignored if _isInputETH 
+     *                      is true as _maxSpend is then equal to msg.value
+     */
+    function exchangeIssue(address _index, uint256 _amount, bool _isInputETH, address _inputToken, uint256 _maxSpend) external payable {
```

Comment:
I believe `exchangeIssue()` shouldn't take index amount as a parameter. But instead, decide the amount of index from the input token amount. Otherwise there might be a scenario if the input token amount might not be sufficient to issue given `_amount` of index tokens, and the whole transaction would revert.